src/main/java/org/kaazing/netx/ws/internal/util/OptimisticReentrantLock.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2014, Kaazing Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.kaazing.netx.ws.internal.util;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+// ### TODO: 1. happens-before relationship should be established between the thread that releases the lock and the next thread
+//              that acquires the lock.
+//           2. Evaluate using Martin's sun.misc.Unsafe based AtomicSequence implementation.
+public class OptimisticReentrantLock {
```

Comment:
I was not planning on implementing fairness as it is not free -- http://www.ibm.com/developerworks/library/j-jtp10264/. 

If we think that we should implement fairness, then we can use `AtomicReferenceArray` with an initial size. If the number of threads increases the initial size, then we can create a new instance of `AtomicReferenceArrray` and copy the contents of the old instance into the new instance.